### PR TITLE
Adding timeseries forecasting example

### PR DESCRIPTION
**Pull request by falaktheoptimist**

### CODE REVIEW

**Comment by PrabhanshuAttri** [comment]
@falaktheoptimist We are already working on this and have a pending PR regarding the same. Feel free to contribute to our branch. Here are the PRs and issue claimed by my team: #102 #117 #106

**Comment by fchollet** [comment]
@falaktheoptimist @PrabhanshuAttri since we have two examples for the same thing, I will review both PRs and ultimately we will merge the one that fits the goals of this example the best (most readable, most didactic, etc).

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+# Extract the required features and target from the dataframe
+feature_list = ["TempAvgF", "month", "day"]
+X_train = data_train[feature_list].values
+y_train = data_train["Target"].values
+X_validation = data_validation[feature_list].values
+y_validation = data_validation["Target"].values
+
+"""
+Use timeseries_dataset_from_array for creating TF.data.dataset object from
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:109]: Use backticks for code keyword. `tf` in lowercase.

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+## Fetch the data
+We will use the [Austin weather data from ](https://www.kaggle.com/grubenm/austin-weather)
+dataset. It provides timeseries of weather related information for Austin.
+
+We will try to predict the Average temperature for next day using temperature
+and date/ month features. The simplicity of this dataset allows us to
+demonstrate the time series use case of LSTM effectively.
+"""
+
+os.system(
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:46]: You can use `keras.utils.get_file` for this

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+import os
+import matplotlib.dates as mdates
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+
+"""
+## Fetch the data
+We will use the [Austin weather data from ](https://www.kaggle.com/grubenm/austin-weather)
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:38]: Add line break above

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+)
+
+# Load the data
+df = pd.read_csv("austin_weather.csv")
+df["Date"] = pd.to_datetime(df["Date"])
+print(df.head(5))
+
+"""
+## Feature generation
+Since weather information is seasonal, features indicating the month
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:57]: Add line break above

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+ax.xaxis.set_major_locator(months)
+ax.xaxis.set_major_formatter(months_fmt)
+
+# rotates and right aligns the x labels, and moves the bottom of the
+# axes up to make room for them
+fig.autofmt_xdate(rotation=90)
+
+"""
+## Input data preparation
+Given weather information till day d, The target is the temperature at day d+1
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:86]: Add line break above, use backticks for code keywords

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+data_df.dropna(inplace=True)
+print(data_df.head())
+
+# Train and validation split
+data_train = data_df[data_df["Date"].dt.year < 2016]
+data_validation = data_df[data_df["Date"].dt.year > 2016]
+
+# Extract the required features and target from the dataframe
+feature_list = ["TempAvgF", "month", "day"]
+X_train = data_train[feature_list].values
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:103]: Use lowercase variable names everywhere.

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+## Input data preparation
+Given weather information till day d, The target is the temperature at day d+1
+The timeseries_dataset_from_array that we'll use below gathers samples
+starting from day `d` to day `d + sequence_length - 1` in a sample.
+So, corresponding target is temperature on day `d + sequence_length`
+We use a sequence_length of 10.
+"""
+data_df = df[["Date", "TempAvgF", "month", "day"]]
+data_df["Target"] = df["TempAvgF"].shift(-10)
+data_df.dropna(inplace=True)
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:94]: Add a comment to explain this

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    break
+inp_arr = inputs.numpy()
+print("Input shape:  {}".format(inp_arr.shape))
+print("Target shape: {}".format(targets.numpy().shape))
+
+print("Sample of input:")
+print(inp_arr[0], "\n\n", inp_arr[1])
+
+"""
+ Note how the last temperature on the last timestamp of sample 2 (11th day)
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:138]: Remove indent

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+ is the target for sample 1 which uses temperatures for day 1-10 as inputs.
+"""
+print("Targets: \n {}".format(targets))
+
+"""
+## Build the LSTM based model
+Single LSTM layer followed by a dense layer.
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:145]: Add line break above

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+"""
+model = keras.Sequential(
+    [
+        layers.Input(shape=(inputs.shape[1], inputs.shape[2])),
+        # Uncomment below line if you want to use intermediate layers.
+        # layers.LSTM(20, return_sequences=True),
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:151]: No need for this

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+ Note how the last temperature on the last timestamp of sample 2 (11th day)
+ is the target for sample 1 which uses temperatures for day 1-10 as inputs.
+"""
+print("Targets: \n {}".format(targets))
+
+"""
+## Build the LSTM based model
+Single LSTM layer followed by a dense layer.
+"""
+model = keras.Sequential(
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:147]: Prefer using the Functional API

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        layers.Dense(1),
+    ]
+)
+model.compile(optimizer=keras.optimizers.Adam(learning_rate=0.01), loss="mse")
+model.summary()
+
+"""
+## Train the model
+"""
+history = model.fit(dataset_train, epochs=200, validation_data=dataset_validation)
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:162]: Add an early stopping and model checkpoint callback

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+"""
+## Setup
+As of writing this tutorial, timeseries_dataset_from_array is available only
+in TF-nightly.
+"""
+
+# tf_nightly needed for timeseries_dataset_from_array
+#! pip uninstall -y tensorflow
+#! pip install tf-nightly
```
Comment (fchollet) [inline examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:24]: You can remove these lines. You just need to say "this example requires TensorFlow 2.3 or higher"

**Comment by fchollet** [comment]
@falaktheoptimist @PrabhanshuAttri:

The example here by @falaktheoptimist is significantly higher quality and simpler than the other one. However, the other one had a few nice visualizations that we could add to this example. Consider merging the other example into this one.

**Comment by flosincapite** [comment]
Hi, @fchollet. The MLH fellows have been working on this example for a while. They indicated as much both in a GitHub issue and in the call for contributions. Can we contribute their PR and then work with @falaktheoptimist to merge the two?

**Comment by fchollet** [comment]
> Hi, @fchollet. The MLH fellows have been working on this example for a while. They indicated as much both in a GitHub issue and in the call for contributions. Can we contribute their PR and then work with @falaktheoptimist to merge the two?

I fully acknowledge that they had precedence. But being first to start working on an example doesn't guarantee you'll get published on keras.io. The key problem is that their example, in its current form, is not at all up to publication standards. It will need very significant work to get there. It is much easier to use the current example in this PR as a starting point instead, since that example is in a better shape. So I suggest we do that.

**Comment by flosincapite** [comment]
Thanks, @fchollet. The Fellows are working really hard to address your comments on their PR. The significant work is underway.

I'll also contact you offline to see how we can avoid something like this in future.

**Comment by fchollet** [comment]
@flosincapite it doesn't matter which PR we use, but since #118 is in a better state, I recommend reusing most the code it contains. You can then just add @falaktheoptimist as a coauthor. Easy.

**Comment by falaktheoptimist** [comment]
Thanks @fchollet for the review.
@PrabhanshuAttri @flosincapite : I just saw the open for contributions page and ended up adding a minimal example. Sorry I did not look at the open PRs page first or would have avoided this overlap/ dual effort.
We can merge both as @fchollet suggested above. You have some pretty cool visualizations.

**Comment by PrabhanshuAttri** [comment]
@fchollet Seems like a good idea. Will fix the issues and add @falaktheoptimist to #123

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py (hunk 13 above)
Comment (falaktheoptimist) [reply examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py]: Done

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py (hunk 6 above)
Comment (falaktheoptimist) [reply examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py]: Done

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py (hunk 2 above)
Comment (falaktheoptimist) [reply examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py]: That's much better. Thanks.

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py (hunk 11 above)
Comment (falaktheoptimist) [reply examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py]: Switched to functional

File: examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py (hunk 7 above)
Comment (falaktheoptimist) [reply examples/timeseries/timeseries_forecasting_with_lstm_for_temperature_prediction.py:96]: Added

**Comment by PrabhanshuAttri** [comment]
@falaktheoptimist I have integrated your code in our branch. Would you mind sharing your GitHub or I can use `falaktheoptimist@users.noreply.github.com` to add you as a contributor?

**Comment by fchollet** [comment]
Closing this PR since we're moving the development of this example to the other PR. Thanks everyone for the quick resolution

**Comment by falaktheoptimist** [comment]
Hi @PrabhanshuAttri :
You can add falaktheoptimist@gmail.com
Are you referring to [this repo?](https://github.com/MLH-Fellowship/keras-io/blob/example/timeseries_dataset_from_array/examples/timeseries/timeseries_weather_forecasting.py?) I didn't see any of my changes there.
From what @fchollet mentioned yesterday, I thought merging was going the other way round. A bit disappointed at how this turned out. 
Anyways, let me know if I can help in any way there.

**Comment by PrabhanshuAttri** [comment]
@falaktheoptimist Yes, that's the repository. I didn't commit anything as I wanted to add you as a co-author. Will add you and push the changes to #123 I have also share the Google Colab notebook with you.

**Pull request closed without merging**